src/Sulu/Component/Content/Repository/ContentRepository.php

```diff
@@ -0,0 +1,636 @@
+        $this->qomFactory = $this->session->getWorkspace()->getQueryManager()->getQOMFactory();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function find($uuid, $locale, $webspaceKey, MappingInterface $mapping, UserInterface $user = null)
+    {
+        $locales = $this->getLocalesByWebspaceKey($webspaceKey);
+        $queryBuilder = $this->getQueryBuilder($locale, $user);
+        $queryBuilder->where(
+            $this->qomFactory->comparison(
+                new PropertyValue('node', 'jcr:uuid'),
+                '=',
+                $this->qomFactory->literal($uuid)
+            )
+        );
+        $this->appendMapping($queryBuilder, $mapping, $locales);
+
+        $rows = $queryBuilder->execute();
+
+        if (count(iterator_to_array($rows->getRows())) !== 1) {
+            throw new ItemNotFoundException();
+        }
+
+        return $this->resolveContent($rows->getRows()->current(), $locale, $webspaceKey, $mapping, $user);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function findByParentUuid(
+        $uuid,
+        $locale,
+        $webspaceKey,
+        MappingInterface $mapping,
+        UserInterface $user = null
+    ) {
+        $path = $this->resolvePathByUuid($uuid);
+
+        $locales = $this->getLocalesByWebspaceKey($webspaceKey);
+        $queryBuilder = $this->getQueryBuilder($locale, $user);
+        $queryBuilder->where($this->qomFactory->childNode('node', $path));
+        $this->appendMapping($queryBuilder, $mapping, $locales);
+
+        return $this->resolveQueryBuilder($queryBuilder, $locale, $webspaceKey, $mapping, $user);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function findByWebspaceRoot($locale, $webspaceKey, MappingInterface $mapping, UserInterface $user = null)
+    {
+        $locales = $this->getLocalesByWebspaceKey($webspaceKey);
+        $queryBuilder = $this->getQueryBuilder($locale, $user);
+        $queryBuilder->where(
+            $this->qomFactory->childNode('node', $this->sessionManager->getContentPath($webspaceKey))
+        );
+        $this->appendMapping($queryBuilder, $mapping, $locales);
+
+        return $this->resolveQueryBuilder($queryBuilder, $locale, $webspaceKey, $mapping, $user);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function findParentsWithSiblingsByUuid(
+        $uuid,
+        $locale,
+        $webspaceKey,
+        MappingInterface $mapping,
+        UserInterface $user = null
+    ) {
+        $contentPath = $this->sessionManager->getContentPath($webspaceKey);
+        $path = $this->resolvePathByUuid($uuid);
+
+        $locales = $this->getLocalesByWebspaceKey($webspaceKey);
+        $queryBuilder = $this->getQueryBuilder($locale, $user)
+            ->orderBy($this->qomFactory->propertyValue('node', 'jcr:path'))
+            ->where($this->qomFactory->childNode('node', $path));
+
+        while ($path !== $contentPath) {
+            $path = dirname($path);
```

Comment:
Feels somehow wrong... but is at the same time the most convenient way to do this I can imagine...